usr/lib/common/mech_aes.c

```diff
@@ -488,6 +488,107 @@ CK_RV aes_ctr_decrypt(STDLL_TokData_t *tokdata,
     return rc;
 }
 
+CK_RV aes_xts_encrypt(STDLL_TokData_t *tokdata,
+                      SESSION *sess,
+                      CK_BBOOL length_only,
+                      ENCR_DECR_CONTEXT *ctx,
+                      CK_BYTE *in_data,
+                      CK_ULONG in_data_len,
+                      CK_BYTE *out_data, CK_ULONG *out_data_len)
+{
+	OBJECT *key = NULL;
+	CK_RV rc;
+
+	if (!sess || !ctx || !out_data_len) {
+		TRACE_ERROR("%s received bad argument(s)\n", __func__);
+		return CKR_FUNCTION_FAILED;
+	}
+	// CKM_DES3_CBC requires the input data to be an integral
+	// multiple of the block size
+	//
+	if (in_data_len % AES_BLOCK_SIZE != 0) {
+		TRACE_ERROR("%s\n", ock_err(ERR_DATA_LEN_RANGE));
+		return CKR_DATA_LEN_RANGE;
+	}
+
+	rc = object_mgr_find_in_map1(tokdata, ctx->key, &key, READ_LOCK);
+	if (rc != CKR_OK) {
+		TRACE_ERROR("Failed to find specified object.\n");
+		return rc;
+	}
+
+	if (length_only == TRUE) {
+		*out_data_len = in_data_len;
+		rc = CKR_OK;
+		goto done;
+	}
+
+	if (*out_data_len < in_data_len) {
+		*out_data_len = in_data_len;
+		TRACE_ERROR("%s\n", ock_err(ERR_BUFFER_TOO_SMALL));
+		rc = CKR_BUFFER_TOO_SMALL;
+		goto done;
+	}
+
+	rc = ckm_aes_xts_encrypt(tokdata, in_data, in_data_len, out_data,
+			out_data_len, ctx->mech.pParameter, key);
+
+	done: object_put(tokdata, key, TRUE);
+	key = NULL;
+
+	return rc;
+}
+
+CK_RV aes_xts_decrypt(STDLL_TokData_t *tokdata,
+                      SESSION *sess,
+                      CK_BBOOL length_only,
+                      ENCR_DECR_CONTEXT *ctx,
+                      CK_BYTE *in_data,
+                      CK_ULONG in_data_len,
+                      CK_BYTE *out_data, CK_ULONG *out_data_len)
+{
+	OBJECT *key = NULL;
+	CK_RV rc;
+
+	if (!sess || !ctx || !out_data_len) {
+		TRACE_ERROR("%s received bad argument(s)\n", __func__);
+		return CKR_FUNCTION_FAILED;
+	}
+	// CKM_DES3_CBC requires the input data to be an integral
+	// multiple of the block size
+	//
+	if (in_data_len % AES_BLOCK_SIZE != 0) {
```

Comment:
As far as I know for XTS the data must be at least a full block, but can be any number of bytes other than that. So for single part operation (this function), the check should be if (in_data_len >= AES_BLOCK_SIZE) {